Fix carryall not removing influence when cancelling land activity

PunkPun opened the pull request:
closes #20402

I think it's a reasonable fix, though it does feel a bit a bit strange as nowhere else is `ChildHasPriority` set during activity runtime.

dnqbob commented:
Umm strange indeed. 

How about making carryall not add influence when fly down to carry?

(1). Carryall won't actually land when carry a unit: it lands on unit's top.
(2). The unit under Carryall always takes up cell, so it will not result in other aircraft can land on the cell.
(3). The unit won't move during carryall flying down, so the cell in (2) won't change during this process

PunkPun commented:
We could remove influence, but I think this also needs fixing in the spirit of keeping landing / takeoff behaviours consistent

dnqbob commented:
> We could remove influence, but I think this also needs fixing in the spirit of keeping landing / takeoff behaviours consistent

Make sense if we look in this way

darkademic commented:
Gave this a quick test and it'll still crash if you order the carryall to pick up a unit, then move away at the very last moment, then try to pick the unit up again (whereas previously you just had to interrupt it at any point during the pickup).

You know it will crash when the carryall moves without rising back up first (whereas if you interrupt before the last moment, it'll rise back up vertically, then move).

dnqbob commented:
> Gave this a quick test and it'll still crash if you order the carryall to pick up a unit, then move away at the very last moment, then try to pick the unit up again (whereas previously you just had to interrupt it at any point during the pickup).
> 
> You know it will crash when the carryall moves without rising back up first (whereas if you interrupt before the last moment, it'll rise back up vertically, then move).

Thanks to @darkademic  I also confirm that there is still problem.
[replay.zip](https://github.com/OpenRA/OpenRA/files/9886712/replay.zip)

You can try this replay or reproduce it like what darkademic said. It is harder to trigger but still crash

PunkPun committed:
Fix carryall not removing influence when cancelling land activity

PunkPun commented:
I found a better fix

PunkPun commented:
We have `MinAirborneAltitude` which description is "Minimum altitude where this aircraft is considered airborne."
and we have `LandAltitude` with its description of "Altitude at which the aircraft considers itself landed."

These 2 values overlap and I recon we need to review them

darkademic commented:
With the updated fix I'm not able to replicate the crash anymore, nor can I make the Carryall glitch sideways before rising back up.

dnqbob commented:
No crash, either

ghost commented:
> We have `MinAirborneAltitude` which description is "Minimum altitude where this aircraft is considered airborne." and we have `LandAltitude` with its description of "Altitude at which the aircraft considers itself landed."
> 
> These 2 values overlap and I recon we need to review them

They don't overlap, they are for different things. `MinAirborneAltitude` is the exposed-to-yaml WW logic, where the "ground-to-air TargetTypes" switch only happens at 200 leptons above ground (800 world units in OpenRA terms) and is required for close-to-ground targetability of aircraft. `LandAltitude` is the Missile's `CloseEnough` exposed-to-yaml rounding counterpart to ensure that the aircraft speed doesn't cause any issues during landing.

They can't be merged, partially due to #14150 (which could mitigate the issues from the former). `LandAltitude` might no longer be necessary if the activity can recognize and notify the other traits that it finished landing/docking.

darkademic commented:
Does this one need rethinking based on the above, or is the fix fine for now?

PunkPun commented:
The fix is not fine I recon. What this assumes is that HasInfluence = Landed / Landing, but if one chooses to not add influence when landing in a case like #20407 then this will induce regressions.

dnqbob commented:
My other new idea is still related to `AddLandInfluence`, but we should give `Land` an option that if this land activity will add land influence, because

1. When land on terrain, we add influence
2. When land on building, we add influence
3. When land on mobile actor, we actually should not add influence. We cannot expect to block the cell behind the mobile actor to land on it.

darkademic commented:
Still able to crash unfortunately. If I just repeatedly click the unit I'm trying to pick up as the carryall descends, it'll crash pretty much immediately.

PunkPun commented:
This should be a more permanent solution

darkademic commented:
More difficult to crash it, but still does when you order the carryall to move at the very last moment when it's picking something up (it does the glitchy sideways movement then crashes on next attempt).

darkademic commented:
Seems to be when you move during the BeforeLoadDelay window. If I set that to zero I can't crash it.

PunkPun commented:
That delay creates a wait activity, and I cannot seem to be able to reproduce the crash

darkademic commented:
> That delay creates a wait activity, and I cannot seem to be able to reproduce the crash

Hmm yea I can't replicate in TS. Can't see any differences that might cause it to crash in CA.

PunkPun committed:
Don't let the player interrupt takeoff after a succesful caryall pickup

PunkPun commented:
I've included another fix, now you won't be able to interrupt the carryall while it's ascending after a successful pickup. I'm still unsure how you were able to crash though

darkademic commented:
Ah I was able to replicate in TS, took quite a few attempts though. Basically just by spam clicking the carryall to move just after it touches the unit it's trying to pick up.

PunkPun commented:
Fixed. 

Wait activity was causing a bit of weirdness by being uninterruptable, I've changed the PickupUnit activity to also be uninterruptable when queueing TakeOff

darkademic commented:
Cool, yep confirmed fixed.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/014157dbd36636446064b24daf8e9e18870710e3...4075a7f2d3c9ba7e0cea7f5ea96fb1f40a325a34)
